### PR TITLE
Creating a workspace item from an external source

### PR DESCRIPTION
**Pull request by benbosman**
This is the REST API development for https://github.com/DSpace/Rest7Contract/pull/83, the changes to workspaceitems.md only

It ensures submitters can create a new entity based on an external source as a workspace item.
This uses a POST on the /api/submission/workspaceitems, but using a URI-list (with the external source)

The item is created as a workspace item, the submitter can further edit the item hereafter.

For code reuse purposes, the ExternalSourceEntryUriListHandler class has been renamed to ExternalSourceEntryItemUriListHandler and made abstract, with the ExternalSourceEntryArchivedItemUriListHandler being used for https://github.com/DSpace/DSpace/pull/2590 and ExternalSourceEntryWorkspaceItemUriListHandler being used for this task

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/external/service/ExternalDataService.java [hunk 1]
```diff
@@ -75,6 +75,7 @@
      * @throws AuthorizeException   If something goes wrong
      * @throws SQLException         If something goes wrong
      */
-    Item createItemFromExternalDataObject(Context context, ExternalDataObject externalDataObject, Collection collection)
+    WorkspaceItem createWorkspaceItemFromExternalDataObject(Context context, ExternalDataObject externalDataObject,
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/external/service/ExternalDataService.java]: The JavaDocs of this method need to be updated.  They currently say:
> This method will create an Item in the given Collection based on the given ExternalDataObject.
> Note that this Item will be Archived

You've modified it to return a WorkspaceItem which is *not* Archived.  I see in the below code you've moved where the Item is archived...that's fine, but we should correct the JavaDocs

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryArchivedItemUriListHandler.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import org.dspace.content.service.CollectionService;
+import org.dspace.content.service.InstallItemService;
+import org.dspace.core.Context;
+import org.dspace.external.service.ExternalDataService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+/**
+ * This class will handle ExternalSourceEntryUriList and it'll create Item objects based on them
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryArchivedItemUriListHandler.java]: We should enhance this description slightly (or enhance JavaDocs in the `handle()` method).  We need to stress here that this `UriListHandler` creates *archived Items* (and as such only an Admin user can use it)

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java [hunk 3]
```diff
@@ -23,21 +23,16 @@
...
 import org.springframework.data.rest.webmvc.ResourceNotFoundException;
-import org.springframework.stereotype.Component;
 import org.springframework.web.bind.annotation.RequestMethod;
 
-/**
- * This class will handle ExternalSourceEntryUriList and it'll create Item objects based on them
- */
-@Component
-public class ExternalSourceEntryUriListHandler implements UriListHandler<Item> {
+public abstract class ExternalSourceEntryItemUriListHandler<T> implements UriListHandler<T> {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java]: Please add Javadoc description of this Abstract class

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java [hunk 4]
```diff
@@ -51,11 +46,9 @@
     private CollectionService collectionService;
 
     private static final Logger log = org.apache.logging.log4j.LogManager
-        .getLogger(ExternalSourceEntryUriListHandler.class);
+        .getLogger(ExternalSourceEntryItemUriListHandler.class);
 
-
-    @Override
-    public boolean supports(List<String> uriList, String method) {
+    public boolean supports(List<String> uriList, String method, Class clazz) {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java]: I think you lost the `@Override` annotation here.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java [hunk 5]
```diff
@@ -68,23 +61,16 @@ public boolean supports(List<String> uriList, String method) {
         return true;
     }
 
-    @Override
-    public boolean validate(Context context, HttpServletRequest request, List<String> uriList,
-                            Class clazz) throws AuthorizeException {
+    public boolean validate(Context context, HttpServletRequest request, List<String> uriList)
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java]: I think you lost the `@Override` annotation here.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java [hunk 6]
```diff
@@ -97,23 +83,19 @@ public boolean validate(Context context, HttpServletRequest request, List<String
         return true;
     }
 
-    @Override
-    public Item handle(Context context, HttpServletRequest request, List<String> uriList)
+    public WorkspaceItem createWorkspaceItem(Context context, HttpServletRequest request, List<String> uriList)
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java]: Please add Javadocs to this method

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryWorkspaceItemUriListHandler.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.List;
+import javax.servlet.http.HttpServletRequest;
+
+import org.dspace.authorize.AuthorizeException;
+import org.dspace.content.WorkspaceItem;
+import org.dspace.core.Context;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ExternalSourceEntryWorkspaceItemUriListHandler
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryWorkspaceItemUriListHandler.java]: Please add JavaDocs

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryWorkspaceItemUriListHandler.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    public boolean validate(Context context, HttpServletRequest request, List<String> uriList)
+        throws AuthorizeException {
+
+        if (!super.validate(context, request, uriList)) {
+            return false;
+        }
+        return true;
+    }
+
+    public WorkspaceItem handle(Context context, HttpServletRequest request, List<String> uriList)
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryWorkspaceItemUriListHandler.java]: Should this have an `@Override`?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/SubmissionService.java [hunk 9]
```diff
@@ -92,7 +92,7 @@
     public WorkspaceItem createWorkspaceItem(Context context, Request request) throws SQLException, AuthorizeException {
         WorkspaceItem wsi = null;
         Collection collection = null;
-        String collectionUUID = request.getHttpServletRequest().getParameter("collection");
+        String collectionUUID = request.getHttpServletRequest().getParameter("owningCollection");
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/SubmissionService.java]: Will this affect the Angular UI submission process?  The REST Contract does say this param should be `owningCollection`, so this change is needed.  However, it seems like we may need a minor PR to the Angular UI to ensure this PR doesn't break submissions.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java [hunk 10]
```diff
@@ -1625,4 +1630,209 @@ public void uploadTest() throws Exception {
...
 
+
+
+
+
+
+
+
+
+    @Test
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java]: There's a large number of blank lines above this test which can be removed.

**Review by tdonohue — COMMENTED** [review]
@benbosman : Gave this a quick code review today (haven't tested yet).  I've added some inline comments below -- most of them are minor requests for JavaDocs.  Overall the code here looks good.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/SubmissionService.java (hunk 9 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/SubmissionService.java]: There's no Angular UI using the collection or owningCollection parameter when creating a workspace item yet

File: dspace-api/src/main/java/org/dspace/external/service/ExternalDataService.java (hunk 1 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/external/service/ExternalDataService.java]: This has been updated

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryArchivedItemUriListHandler.java (hunk 2 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryArchivedItemUriListHandler.java]: This has been updated

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java (hunk 3 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java]: This has been updated

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java (hunk 4 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java]: This has been updated

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java (hunk 5 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java]: This has been updated

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java (hunk 6 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryItemUriListHandler.java]: This has been updated

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryWorkspaceItemUriListHandler.java (hunk 7 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryWorkspaceItemUriListHandler.java]: This has been updated

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryWorkspaceItemUriListHandler.java (hunk 8 above)
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/handler/ExternalSourceEntryWorkspaceItemUriListHandler.java]: Yes, this has been updated

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java (hunk 10 above)
Comment (benbosman) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkspaceItemRestRepositoryIT.java]: This has been updated

**Comment by benbosman** [comment]
Thanks @tdonohue for your review, Raf has adjusted everything you requested

**Review by paulo-graca — APPROVED** [review]
The PR works as expected and the code changes also look ok to me.
One thing related with the angular interface. Since for workspace items it's only prepared for items that have a title (dc.title), if we create a new item Person from Orcid, it will be difficult to identify which entry is it. As an usability improvement suggestion, all Entities could have a dc.title (as a virtual metadata).

**Comment by benbosman** [comment]
Thanks @paulo-graca for your review.
It's a known issue that MyDSpace UI doesn't use the item list component yet for rendering the list of items. Once that's solved, this complexity will be solved automatically.

**Review by tdonohue — APPROVED** [review]
👍 Gave this another code review, and it looks correct to me. Thanks @benbosman and @Raf-atmire

**Comment by tdonohue** [comment]
Merging, as this is now at +2

**Pull request merged**